src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -0,0 +1,161 @@
+#include "ray/core_worker/transport/direct_task_transport.h"
+
+namespace ray {
+
+void DoInlineObjectValue(const ObjectID &obj_id, std::shared_ptr<RayObject> value,
+                         const TaskSpecification &task) {
+  auto &msg = task.GetMutableMessage();
+  bool found = false;
+  for (size_t i = 0; i < task.NumArgs(); i++) {
+    auto count = task.ArgIdCount(i);
+    if (count > 0) {
+      const auto &id = task.ArgId(i, 0);
+      if (id == obj_id) {
+        auto *mutable_arg = msg.mutable_args(i);
+        mutable_arg->clear_object_ids();
+        if (value->HasData()) {
+          const auto &data = value->GetData();
+          mutable_arg->set_data(data->Data(), data->Size());
+        }
+        if (value->HasMetadata()) {
+          const auto &metadata = value->GetMetadata();
+          mutable_arg->set_metadata(metadata->Data(), metadata->Size());
+        }
+        found = true;
+      }
+    }
+  }
+  RAY_CHECK(found) << "obj id " << obj_id << " not found";
+}
+
+void LocalDependencyResolver::ResolveDependencies(const TaskSpecification &task,
+                                                  std::function<void()> on_complete) {
+  absl::flat_hash_set<ObjectID> local_dependencies;
+  for (size_t i = 0; i < task.NumArgs(); i++) {
+    auto count = task.ArgIdCount(i);
+    if (count > 0) {
+      RAY_CHECK(count <= 1) << "multi args not implemented";
+      const auto &id = task.ArgId(i, 0);
+      if (id.IsDirectActorType()) {
+        local_dependencies.insert(id);
+      }
+    }
+  }
+  if (local_dependencies.empty()) {
+    on_complete();
+    return;
+  }
+
+  TaskState *state = new TaskState{task, std::move(local_dependencies)};
+
+  if (!state->local_dependencies.empty()) {
```

Comment:
shouldn't we still call `on_complete` in this case? also is this condition even meaningful?